contracts/transfer/src/state.rs

```diff
@@ -182,26 +292,113 @@ impl TransferState {
+                &call.fn_args,
+            );
+        }
+
+        result
+    }
+
+    /// Spends the amount available to the moonlight transaction, and executes
+    /// the contract call if present. It performs all checks necessary to ensure
+    /// the transaction is valid - signature check, available funds, etc...
+    ///
+    /// This will emplace the deposit in the state, if it exists - making it
+    /// available for any contracts called.
+    ///
+    /// [`refund`] **must** be called if this function succeeds, otherwise we
+    /// will have an inconsistent state.
+    ///
+    /// # Panics
+    /// Any failure in the checks performed in processing the transaction will
+    /// result in a panic. The contract expects the environment to roll back any
+    /// change in state.
+    ///
+    /// [`refund`]: [`TransferState::refund`]
+    fn spend_and_execute_moonlight(
+        &mut self,
+        tx: MoonlightTransaction,
+    ) -> Result<Vec<u8>, ContractError> {
+        transitory::put_transaction(tx);
+        let tx = transitory::unwrap_moonlight_tx();
+
+        // check the signature is valid and made by `from`
+        let payload = tx.payload();
+        let signature = *tx.signature();
+
+        let from = payload.from;
+        let digest = tx.signature_message();
+
+        if !rusk_abi::verify_bls(digest, from, signature) {
+            panic!("Invalid signature!");
+        }
+
+        // check `from` has the funds necessary to supress the total value
+        // available in this transaction, and that the `nonce` is higher than
+        // the currently held number. if these conditions are violated we panic
+        // since the transaction is invalid - either because the account doesn't
+        // have (enough) funds, or because they're possibly trying to reuse a
+        // previously used signature (i.e. a replay attack).
+        //
+        // Afterwards be simply deduct the total amount of the transaction from
+        // the balance, increment the nonce, and rely on `refund` to be called
+        // after a successful exit.
+        let from_bytes = from.to_bytes(); // TODO: this is expensive. maybe we should address the
+                                          //       fact that `AccountPublicKey` doesn't impl `Ord`
+                                          //       so we can just use it directly as a key in the
+                                          //       `BTreeMap`
+
+        // the total value carried by a transaction is the sum of the value, the
+        // deposit, and gas limit * gas_price.
+        let total_value = payload.value
+            + payload.deposit
+            + payload.gas_limit * payload.gas_price;
+
+        match self.accounts.get_mut(&from_bytes) {
+            Some(account) => {
+                if total_value > account.balance {
+                    panic!("Account doesn't have enough funds");
+                }
+                if payload.nonce <= account.nonce {
```

Comment:
Perhaps this would be better. However, I have to point out that this also restricts users from sending multiple transaction, with some of them possibly failing. Another idea would be to "wrap around" when the end is reached, at the risk of repeats.

Reply:
We should probably discuss this somewhere. Other (successful) chains also use this exact mechanism, so it might not be that bad. `u64::MAX` is pretty big :blush:

Reply:
The fact that other chains use it, doesn't make it right :-p
I know the value is very big, but a bugged implementation could easily "reach it". :)

WRT to sending multiple txs, I think this change would not have any effect on that. It just restricts the client to always use last_value+1, instead of an arbitrary bigger value. But maybe I'm missing your point.

That said, I agree with discussing it somewhere else.